### ISSUE TITLE
Eval fails for singleton types

### ISSUE THREAD

**Issue by scabug** [scala/bug#6465]
```scala
import scala.reflect.runtime.universe._
import scala.tools.reflect.Eval

object Test extends App {
  object i
  reify{val x: i.type = i; x} .eval
}
```

{code}
C:\Projects\Kepler\sandbox @ topic/reflection>myke run Test.scala
scala.tools.reflect.ToolBoxError: reflective compilation has failed:

type mismatch;
 found   : i.type
 required: AnyRef
Note that i extends Any, not AnyRef.
Such types can participate in value classes, but instances
cannot appear in singleton types or in reference comparisons.
        at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$ToolBoxGlobal.throwIfErrors(ToolBoxFactory.scala:318)
        at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$ToolBoxGlobal.compile(ToolBoxFactory.scala:251)
        at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl.compile(ToolBoxFactory.scala:415)
        at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl.eval(ToolBoxFactory.scala:418)
        at scala.tools.reflect.Eval.eval(package.scala:116)
        at Test$delayedInit$body.apply(Test.scala:6)
        at scala.Function0$class.apply$mcV$sp(Function0.scala:40)
        at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
        at scala.App$$anonfun$main$1.apply(App.scala:71)
        at scala.App$$anonfun$main$1.apply(App.scala:71)
        at scala.collection.immutable.List.foreach(List.scala:309)
        at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32)
        at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
        at scala.App$class.main(App.scala:71)
        at Test$.main(Test.scala:4)
        at Test.main(Test.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
        at java.lang.reflect.Method.invoke(Method.java:597)
        at scala.tools.nsc.util.ScalaClassLoader$$anonfun$run$1.apply(ScalaClassLoader.scala:71)
        at scala.tools.nsc.util.ScalaClassLoader$class.asContext(ScalaClassLoader.scala:31)
        at scala.tools.nsc.util.ScalaClassLoader$URLClassLoader.asContext(ScalaClassLoader.scala:139)
        at scala.tools.nsc.util.ScalaClassLoader$class.run(ScalaClassLoader.scala:71)
        at scala.tools.nsc.util.ScalaClassLoader$URLClassLoader.run(ScalaClassLoader.scala:139)
        at scala.tools.nsc.CommonRunner$class.run(ObjectRunner.scala:28)
        at scala.tools.nsc.ObjectRunner$.run(ObjectRunner.scala:45)
        at scala.tools.nsc.CommonRunner$class.runAndCatch(ObjectRunner.scala:35)
        at scala.tools.nsc.ObjectRunner$.runAndCatch(ObjectRunner.scala:45)
        at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:74)
        at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:96)
        at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:105)
        at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala){code}

**Comment by SethTisue** [comment]
weird... works in 2.12.18, but then in 2.13.11 it fails with a different error:

```
scala.tools.reflect.ToolBoxError: reflective compilation has failed:

Internal error: unable to find the outer accessor symbol of object __wrapper$1$b8c7f0eeec764414aae123cf927dff71
Internal error: unable to find the outer accessor symbol of <none>
  at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$ToolBoxGlobal.throwIfErrors(ToolBoxFactory.scala:332)
  at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$ToolBoxGlobal.wrapInPackageAndCompile(ToolBoxFactory.scala:214)
  at scala.tools.reflect.ToolBoxFactory$ToolBoxImpl$ToolBoxGlobal.compile(ToolBoxFactory.scala:268)
```

🤷

**Comment by som-snytt** [comment]
Works for me.
```
import scala.reflect.runtime.universe._
import scala.tools.reflect.Eval

object Test extends App {
  object i {
    override def toString = "I am i"
  }
  println {
    reify {
      val x: i.type = i
      x
    }.eval
  }
}
```
then
```
➜  snips scalac -d /tmp t6465.scala && scala -cp /tmp Test
I am i
➜  snips scala t6465.scala
I am i
➜  snips
```
I was nerdsniped into verifying that it was fixed in 2.12.14. What is the word for procrastinator-sniped?

Pro-tip: when trying out old versions, beware `-nc` compilation daemon.

**Comment by SethTisue** [comment]
ah, I was lazy and didn't question whether the failure I saw might have been REPL-only
